### PR TITLE
879/dataframe

### LINKED ISSUE #879: Create `DataFrameSource`, `DataFrameAdapter` & `PandasDriver`
### Kind of request

Changing existing functionality

### Enhancement Description

Because of the current simplicity of the drivers for a DataFrame, I would tackle these in one issue

```python
lf.driver in ["csv"]:
    df = pd.read_csv(path, **kwargs)
elif self.driver == "parquet":
    _ = kwargs.pop("index_col", None)
    df = pd.read_parquet(path, **kwargs)
elif self.driver in ["xls", "xlsx", "excel"]:
    df = pd.read_excel(path, engine="openpyxl", **kwargs)
elif self.driver in ["fwf"]:
    df = pd.read_fwf(path, **kwargs)
```

### Use case

Read in DataFrames

### Additional Context

Should also create `DataFrameSource` and `DataFrameAdapter`.

### PR DESCRIPTION
**Pull request by Jaapel**
## Issue addressed
Closes #879
## Explanation
- Created a new `DataFrameSource`, `DataFrameAdapter` and `DataFrameDriver` to support (de)serializing, transforming, and I/O with pandas `DataFrame`s.
- Also updated some tests in using `RasterDataset`s, as I noticed some mistakes from my previous commits.
- Removed the tests basically only testing the `options` dot accessor in a convoluted way.

## Checklist
- [x] Updated tests or added new tests
- [x] Branch is up to date with `v1`
- [x] Tests & pre-commit hooks pass
- [x] Updated documentation if needed
- [x] Updated changelog.rst if needed

### CODE REVIEW

File: tests/drivers/test_rasterdataset_driver.py
Comment (savente93) [inline tests/drivers/test_rasterdataset_driver.py:1]: Why was this deleted? I couldn't find the code anywhere else. Please mention on the PR description why it was removed or transferred (which ever is appropriate)

File: tests/drivers/test_geodataframe_driver.py
Comment (savente93) [inline tests/drivers/test_geodataframe_driver.py:1]: Why was this deleted? I couldn't find the code anywhere else. Please mention on the PR description why it was removed or transferred (which ever is appropriate)

File: tests/drivers/test_pandas_driver.py
Comment (savente93) [inline tests/drivers/test_pandas_driver.py:1]: Very nice tests!

File: hydromt/drivers/pandas_driver.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+        Not all drivers should have a write function, so this method is not
+        abstract.
+
+        args:
+        """
+        if isinstance(path, str):
+            extension: str = path.split(".")[-1]
+        elif isinstance(path, Path):
+            extension: str = path.suffix[1:]
```
Comment (savente93) [inline hydromt/drivers/pandas_driver.py:74]:
```suggestion
           # suffix includes the '.' which we don't want
           extension: str = path.suffix[1:]
```

File: hydromt/drivers/pandas_driver.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            return pd.read_parquet(uri, columns=variables, **self.options)
+        elif extension in ["xls", "xlsx"]:
+            return pd.read_excel(
+                uri, usecols=variables, engine="openpyxl", **self.options
+            )
+        elif extension in ["fwf"]:
+            warn_on_unused_kwargs(
+                self.__class__.__name__, {"variables": variables}, logger
+            )
+            return pd.read_fwf(uri, **self.options)
```
Comment (savente93) [inline hydromt/drivers/pandas_driver.py:53]: seems like this implementation can read `fwf` but not write it. Is that correct? I'm not familiar with the format. Do we need it at all?

**Review by savente93 — COMMENTED** [review]
Very nice work Jaap, the tests are structured very well. Just a few small mistakes, and please update the PR description. I'll be happy to approve once those are taken care of. It's fine to not update the migration guide for this one.

File: hydromt/drivers/pandas_driver.py (hunk 2 above)
Comment (Jaapel) [reply hydromt/drivers/pandas_driver.py:53]: Yes, pandas does not write to fwf: https://pandas.pydata.org/docs/user_guide/io.html

File: hydromt/data_source/dataframe.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+        if driver_override:
+            driver: DataFrameDriver = driver_override
+        else:
+            # use local filesystem
+            driver: DataFrameDriver = self.driver.model_copy(
+                update={"filesystem": filesystem("local")}
+            )
+        update.update({"driver": driver})
+
```
[the remark this answers is not in the document]
Comment (Jaapel) [reply hydromt/data_source/dataframe.py:101]: Not yet unfortately, we still need to decide which drivers are default. If we have, the implementation is trivial.

File: tests/drivers/test_rasterdataset_driver.py
Comment (Jaapel) [reply tests/drivers/test_rasterdataset_driver.py:1]: I removed some unused and unnecessary fixtures.

Comment (Jaapel) [reply tests/drivers/test_rasterdataset_driver.py:1]: Ok in this case I was basically just testing whether `options` is a thing, which I think is a bad test haha

File: hydromt/data_source/dataframe.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            file_path,
+            df,
+            **kwargs,
+        )
+
+        return self.model_copy(update=update)
+
+    def to_stac_catalog(
+        self,
+        on_error: ErrorHandleMethod = ErrorHandleMethod.COERCE,
```
[the remark this answers is not in the document]
Comment (Jaapel) [reply hydromt/data_source/dataframe.py:112]: Yeah just pick one indeed.

**Review by savente93 — APPROVED** [review]
Nice!
